modules/tinymce/src/themes/silver/main/ts/ui/context/ContextToolbar.ts

```diff
@@ -0,0 +1,276 @@
+import { UiFactoryBackstage, UiFactoryBackstageProviders } from '../../backstage/Backstage';
+import { renderToolbar } from '../toolbar/CommonToolbar';
+import { identifyButtons } from '../toolbar/Integration';
+import { hideContextToolbarEvent, showContextToolbarEvent } from './ContextEditorEvents';
+import { ContextForm } from './ContextForm';
+import * as ContextToolbarAnchor from './ContextToolbarAnchor';
+import * as ContextToolbarBounds from './ContextToolbarBounds';
+import * as ToolbarLookup from './ContextToolbarLookup';
+import * as ToolbarScopes from './ContextToolbarScopes';
+import { forwardSlideEvent, renderContextToolbar } from './ContextUi';
+
+type ScopedToolbars = ToolbarScopes.ScopedToolbars;
+
+export type ContextSpecType = InlineContent.ContextToolbarSpec | InlineContent.ContextFormSpec;
+export type ContextType = InlineContent.ContextToolbar | InlineContent.ContextForm;
+
+type ContextToolbarButtonType = Toolbar.ToolbarButtonSpec | Toolbar.ToolbarMenuButtonSpec | Toolbar.ToolbarSplitButtonSpec | Toolbar.ToolbarToggleButtonSpec | Toolbar.GroupToolbarButtonSpec;
+
+interface Extras {
+  readonly backstage: UiFactoryBackstage;
+}
+
+const enum TriggerCause {
+  Reposition,
+  NewAnchor
+}
+
+const register = (editor: Editor, registryContextToolbars: Record<string, ContextSpecType>, sink: AlloyComponent, extras: Extras) => {
+  const backstage = extras.backstage;
+  const sharedBackstage = backstage.shared;
+  const isTouch = PlatformDetection.detect().deviceType.isTouch;
+
+  const lastElement = Singleton.value<SugarElement<Element>>();
+  const lastPosition = Singleton.value<Bounds>();
+  const lastTrigger = Singleton.value<TriggerCause>();
+  const lastBounds = Singleton.value<Bounds>();
+
+  const contextbar = GuiFactory.build(
+    renderContextToolbar({
+      sink,
+      onEscape: () => {
+        editor.focus();
+        return Optional.some(true);
+      }
+    })
+  );
+
+  const getBounds = () => {
+    const bounds = ContextToolbarBounds.getContextToolbarBounds(editor, sharedBackstage);
+    lastBounds.set(bounds);
+    return bounds;
+  };
+
+  const shouldContextToolbarHide = (): boolean => {
+    // If a mobile context menu is open, don't launch else they'll probably overlap. For android, specifically.
+    if (isTouch() && backstage.isContextMenuOpen()) {
+      return true;
+    }
+
+    const lastElementBounds = ContextToolbarBounds.getAnchorElementBounds(editor, lastElement.get());
+    const contextToolbarBounds = getBounds();
+
+    // If the element bound isn't overlapping with the context toolbar bounds, the context toolbar should hide
+    return !ContextToolbarBounds.isVerticalOverlap(lastElementBounds, contextToolbarBounds);
+  };
+
+  const close = () => {
+    lastElement.clear();
+    lastPosition.clear();
+    lastTrigger.clear();
+    lastBounds.clear();
+    InlineView.hide(contextbar);
+  };
+
+  const hideOrRepositionIfNecessary = () => {
```

Comment:
I wonder if this should be `hideOrReposition`

I took the "if necessary" to mean it might do neither but it appears to always do one or the other.

Reply:
It will do neither if the context toolbar is closed (hence the isOpen check) and that's why it has `IfNecessary`.